consul/pool_discovery.go

```diff
@@ -0,0 +1,207 @@
+			}
+			req.UpdateCh <- pool
+		}
+	}
+}
+
+// RegisterAgent satisfies the RegisterAgent function on the PoolDiscovery
+// interface.
+func (c *Consul) RegisterAgent(ctx context.Context) error {
+
+	agentID, err := ha.GenerateAgentID()
+	if err != nil {
+		return err
+	}
+
+	// Set important information used when registering this agent.
+	c.agentID = agentID
+	c.serviceID = c.generateServiceID()
+
+	if err := c.registerAgent(); err != nil {
+		return err
+	}
+
+	// Start a goroutine to deregister me on context cancellation. This is best
+	// effort only. We should not remove the ID file if it exists. When running
+	// on metal, shutdown of the autoscaler could be due to maintenance and we
+	// want to come back up with the same ID; like how Nomad does.
+	go func() {
+		<-ctx.Done()
+		_ = c.deregisterAgent()
+		c.poolDiscoveryDoneCh <- struct{}{}
+	}()
+
+	return nil
+}
+
+// Reload satisfies the Reload function on the PoolDiscovery interface.
+func (c *Consul) Reload() { c.poolDiscoveryReloadCh <- struct{}{} }
+
+// WaitForExit satisfies the WaitForExit function on the PoolDiscovery
+// interface.
+func (c *Consul) WaitForExit() {
+	select {
+	case <-time.After(shutdownTimeout):
+	case <-c.poolDiscoveryDoneCh:
+	}
+}
+
+// registerAgent performs the Autoscaler agent registration within Consul along
+// with a check. This check is a basic TCP check targeting the HTTP API.
+func (c *Consul) registerAgent() error {
+
+	// Generate a service registration for this agent. The meta provides an
+	// insight to operators as to where it came from, but as yet we don't have
+	// the flashy label Nomad has.
+	serviceReg := api.AgentServiceRegistration{
+		ID:      c.serviceID,
+		Name:    "nomad-autoscaler",
+		Tags:    []string{autoscalerConsulServiceTag},
+		Address: c.agentAddr,
+		Port:    c.agentPort,
+		Meta:    map[string]string{"external-source": "nomad-autoscaler"},
+	}
+
+	// Generate a check registration object, ensuring the ID is unique.
+	checkReg := api.AgentCheckRegistration{
+		Name:      "nomad-autoscaler-agent",
+		ServiceID: serviceReg.ID,
+		AgentServiceCheck: api.AgentServiceCheck{
+			Interval: "10s",
+			TCP:      net.JoinHostPort(c.agentAddr, strconv.Itoa(c.agentPort)),
+		},
+	}
+	checkReg.ID = generateCheckID(checkReg, c.agentID)
+
+	c.clientLock.RLock()
+	defer c.clientLock.RUnlock()
+
+	// Retry and backoff when attempting to register the service within Consul.
+	serviceFunc := func() error { return c.client.Agent().ServiceRegister(&serviceReg) }
+
+	if err := backoff.Retry(serviceFunc, generateRetryBackoff()); err != nil {
+		return fmt.Errorf("failed to register service: %v", err)
+	}
+
+	// Retry and backoff when attempting to register the service check within
+	// Consul.
+	checkFunc := func() error { return c.client.Agent().CheckRegister(&checkReg) }
+
+	if err := backoff.Retry(checkFunc, generateRetryBackoff()); err != nil {
+		return fmt.Errorf("failed to register service check: %v", err)
+	}
```

Comment:
Probably a minor thing, but these backoffs would hold a read lock for the Consul client even if not using it right? Maybe the lock and unluck should happen inside each function?

Though I don't see where the lock is being acquired for writing.